pythclient/calendar.py

```diff
@@ -185,7 +233,126 @@ def get_next_market_close(asset_type: str, dt: datetime.datetime) -> int:
             next_market_close += datetime.timedelta(days=1)
         while is_market_open(asset_type, next_market_close):
             next_market_close += datetime.timedelta(days=1)
+    elif asset_type == "rates":
+        if dt.date() in NYSE_EARLY_HOLIDAYS:
+            if time < NYSE_EARLY_CLOSE:
+                next_market_close = dt.replace(
+                    hour=NYSE_EARLY_CLOSE.hour,
+                    minute=NYSE_EARLY_CLOSE.minute,
+                    second=0,
+                    microsecond=0,
+                )
+            else:
+                next_market_close = dt.replace(
+                    hour=RATES_CLOSE.hour,
+                    minute=RATES_CLOSE.minute,
+                    second=0,
+                    microsecond=0,
+                )
+                next_market_close += datetime.timedelta(days=1)
+        elif dt.date() in NYSE_HOLIDAYS:
+            next_market_open = get_next_market_open(asset_type, dt)
+            next_market_open_date = (
+                datetime.datetime.fromtimestamp(next_market_open)
+                .astimezone(NY_TZ)
+                .date()
+            )
+            if next_market_open_date in NYSE_EARLY_HOLIDAYS:
+                next_market_close = (
+                    datetime.datetime.fromtimestamp(next_market_open)
+                    .astimezone(NY_TZ)
+                    .replace(
+                        hour=NYSE_EARLY_CLOSE.hour,
+                        minute=NYSE_EARLY_CLOSE.minute,
+                        second=0,
+                        microsecond=0,
+                    )
+                )
+            else:
+                next_market_close = (
+                    datetime.datetime.fromtimestamp(next_market_open)
+                    .astimezone(NY_TZ)
+                    .replace(
+                        hour=RATES_CLOSE.hour,
+                        minute=RATES_CLOSE.minute,
+                        second=0,
+                        microsecond=0,
+                    )
+                )
+        else:
+            next_market_close = dt.replace(
+                hour=RATES_CLOSE.hour,
+                minute=RATES_CLOSE.minute,
+                second=0,
+                microsecond=0,
+            )
+            if time >= RATES_CLOSE:
+                next_market_close += datetime.timedelta(days=1)
+
+        # while next_market_close.date() is in NYSE_HOLIDAYS or weekend, add 1 day
+        while (
+            next_market_close.date() in NYSE_HOLIDAYS
+            or next_market_close.weekday() >= 5
+        ):
+            next_market_close += datetime.timedelta(days=1)
     else:  # crypto markets never close
         return None
 
     return int(next_market_close.timestamp())
+
+
+def get_market_hours_clause(time_column_name):
```

Comment:
Where is this gonna be used?

Reply:
data workflows for calculating publishers uptime